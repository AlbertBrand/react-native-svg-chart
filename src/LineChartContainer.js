/* @flow */
import React, { Component } from 'react';
import { View } from 'react-native';
import { minBy, maxBy } from 'paths-js/ops';
import linear from 'paths-js/linear';
import LineChart from './LineChart';
import calculateStepSize from './calculateStepSize';

export type TickPoints = Array<number>;
export type Point = [number, number, any];
export type Points = Array<Point>;
export type PointScale = (point: Point) => Point;
export type NumberScale = (value: number) => number;
export type Label = (value: number, origValue: any) => string;
export type Data = Array<any>;
export type TextStyle = {
  color: string,
  fontFamily: string,
  fontSize: number,
};
export type Accessor = (value: any, idx: number) => number;
type Event = {
  nativeEvent: {
    layout: {
      width:number,
      height: number
    }
  }
};
type Props = {
  data: Data,
  options: {
    dashLineColor: string,
    dataLineColor: string,
    textStyle: TextStyle,
    xAccessor: Accessor,
    yAccessor: Accessor,
    xLabel: Label,
    yLabel: Label,
  },
  style?: any
};

const MARGIN_TOP = 10;
const MARGIN_RIGHT = 50;
const MARGIN_BOTTOM = 40;
const MARGIN_LEFT = 20;

const MIN_TICK_COUNT = 5;

export default class LineChartContainer extends Component {
  state: {
    layout?: {
      width: number,
      height: number,
    }
  } = {};
  props: Props;
  handle: number;

  renderChart(width: number, height: number) {
    // calculate plot size (area of chart without axes)
    const plotWidth = width - MARGIN_LEFT - MARGIN_RIGHT;
    const plotHeight = height - MARGIN_TOP - MARGIN_BOTTOM;

    // access provided data and transform into chart data
    const {
      xAccessor,
      yAccessor,
    } = this.props.options;
    const transform = (data, idx) => [xAccessor(data, idx), yAccessor(data, idx), data];
    const chartPoints = this.props.data.map(transform);

    // determine min and max
    const xMin = minBy(chartPoints, point => point[0]);
    const xMax = maxBy(chartPoints, point => point[0]);
    const yMin = minBy(chartPoints, point => point[1]);
    const yMax = maxBy(chartPoints, point => point[1]);

    // calculate placement of ticks on y-axis
    const yTickStep = calculateStepSize(yMax - yMin, MIN_TICK_COUNT);
    const yTickMin = Math.floor(yMin / yTickStep) * yTickStep;
    const yTickMax = Math.ceil(yMax / yTickStep) * yTickStep;
    const yTickPoints:TickPoints = [];
    for (let i = yTickMin; i < yTickMax + yTickStep; i += yTickStep) {
      yTickPoints.push(i);
    }

    // create scaling functions for plot width and height
    const xScale = linear([xMin, xMax], [0, plotWidth]);
    const yScale = linear([yTickMin, yTickMax], [plotHeight, 0]);
    const scale = ([x, y]) => [xScale(x), yScale(y)];

    const chartProps = {
      ...this.props.options,
      width,
      height,
      chartPoints,
      yTickPoints,
      xMin,
      xMax,
      xScale,
      xOffset: MARGIN_LEFT,
      yMin: yTickMin,
      yMax: yTickMax,
      yOffset: MARGIN_TOP,
      yScale,
      scale,
    };

    return (
      <LineChart {...chartProps} />
    );
  }

  render() {
    return (
      <View
        onLayout={(e:Event) => this.setState(e.nativeEvent)}
        style={this.props.style}
      >
        {this.state.layout ?
          this.renderChart(this.state.layout.width, this.state.layout.height) : null}
      </View>
    );
  }
}
