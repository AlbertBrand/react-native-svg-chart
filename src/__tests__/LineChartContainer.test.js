// @flow
// eslint-env jest
import 'react-native';
import renderer from 'react-test-renderer';
import React from 'react';
import LineChartContainer from '../LineChartContainer';

jest.mock('../LineChart', () => 'LineChart');

describe('LineChartContainer', () => {
  const xAccessor = value => value.x;
  const yAccessor = value => value.y * 2;
  const xLabel = value => String(value);
  const yLabel = value => String(value);
  const data = [
    { x: 1, y: 2.4 },
    { x: 2, y: 4.9 },
    { x: 4, y: 3.1 },
  ];
  const options = {
    width: 300,
    height: 200,
    dashLineColor: 'gray',
    dataLineColor: 'blue',
    textStyle: {
      color: 'black',
      fontFamily: 'arial',
      fontSize: 10,
    },
    xAccessor,
    yAccessor,
    xLabel,
    yLabel,
  };

  it('renders empty view', () => {
    const tree = renderer.create(
      <LineChartContainer data={data} options={options} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders chart', () => {
    const chartRenderer = renderer.create(
      <LineChartContainer data={data} options={options} />,
    );
    let tree = chartRenderer.toJSON();
    tree.props.onLayout({ nativeEvent: { layout: { width: 300, height: 200 } } });
    tree = chartRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
