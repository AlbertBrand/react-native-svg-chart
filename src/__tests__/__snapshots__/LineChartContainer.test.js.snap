exports[`LineChartContainer renders chart 1`] = `
<View
  onLayout={[Function]}
  style={undefined}>
  <LineChart
    chartPoints={
      Array [
        Array [
          1,
          4.8,
          Object {
            "x": 1,
            "y": 2.4,
          },
        ],
        Array [
          2,
          9.8,
          Object {
            "x": 2,
            "y": 4.9,
          },
        ],
        Array [
          4,
          6.2,
          Object {
            "x": 4,
            "y": 3.1,
          },
        ],
      ]
    }
    dashLineColor="gray"
    dataLineColor="blue"
    height={200}
    scale={[Function]}
    textStyle={
      Object {
        "color": "black",
        "fontFamily": "arial",
        "fontSize": 10,
      }
    }
    width={300}
    xAccessor={[Function]}
    xLabel={[Function]}
    xMax={4}
    xMin={1}
    xOffset={20}
    xScale={[Function]}
    yAccessor={[Function]}
    yLabel={[Function]}
    yMax={10}
    yMin={4}
    yOffset={10}
    yScale={[Function]}
    yTickPoints={
      Array [
        4,
        5,
        6,
        7,
        8,
        9,
        10,
      ]
    } />
</View>
`;

exports[`LineChartContainer renders empty view 1`] = `
<View
  onLayout={[Function]}
  style={undefined} />
`;
